dlt/pipeline/pipeline.py

```diff
@@ -90,48 +90,52 @@ def _wrap(self: "Pipeline", *args: Any, **kwargs: Any) -> Any:
-        trace: PipelineTrace = self._trace
-        trace_step: PipelineStepTrace = None
-        step_info: Any = None
-        is_new_trace = self._trace is None and self.config.enable_runtime_trace
+    def decorator(f: TFun) -> TFun:
 
-        # create a new trace if we enter a traced function and there's no current trace
-        if is_new_trace:
-            self._trace = trace = start_trace(cast(TPipelineStep, f.__name__), self)
+        @wraps(f)
+        def _wrap(self: "Pipeline", *args: Any, **kwargs: Any) -> Any:
+            trace: PipelineTrace = self._trace
+            trace_step: PipelineStepTrace = None
+            step_info: Any = None
+            is_new_trace = self._trace is None and self.config.enable_runtime_trace
 
-        try:
-            # start a trace step for wrapped function
-            if trace:
-                trace_step = start_trace_step(trace, cast(TPipelineStep, f.__name__), self)
+            # create a new trace if we enter a traced function and there's no current trace
+            if is_new_trace:
+                self._trace = trace = start_trace(cast(TPipelineStep, f.__name__), self)
 
-            step_info = f(self, *args, **kwargs)
-            return step_info
-        except Exception as ex:
-            step_info = ex  # step info is an exception
-            raise
-        finally:
             try:
-                if trace_step:
-                    # if there was a step, finish it
-                    end_trace_step(self._trace, trace_step, self, step_info)
-                if is_new_trace:
-                    assert trace is self._trace, f"Messed up trace reference {id(self._trace)} vs {id(trace)}"
-                    end_trace(trace, self, self._pipeline_storage.storage_path)
+                # start a trace step for wrapped function
+                if trace:
+                    trace_step = start_trace_step(trace, cast(TPipelineStep, f.__name__), self)
+
+                step_info = f(self, *args, **kwargs)
+                return step_info
+            except Exception as ex:
+                step_info = ex  # step info is an exception
+                raise
             finally:
-                # always end trace
-                if is_new_trace:
-                    assert self._trace == trace, f"Messed up trace reference {id(self._trace)} vs {id(trace)}"
-                    # if we end new trace that had only 1 step, add it to previous trace
-                    # this way we combine several separate calls to extract, normalize, load as single trace
-                    # the trace of "run" has many steps and will not be merged
-                    self._last_trace = merge_traces(self._last_trace, trace)
-                    self._trace = None
+                try:
+                    if trace_step:
+                        # if there was a step, finish it
+                        end_trace_step(self._trace, trace_step, self, step_info, send_state)
```

Comment:
OK we can send it in `end_trace_step` or here... probably you are right - we can make it trace module specific